Add agent framework/throttling/hidden model/OS assistant and update conversational search documentation (#6354)
opensearch-project/documentation-website: _ml-commons-plugin/api/controller-apis/create-controller.md

---
layout: default
title: Create controller
parent: Controller APIs
grand_parent: ML Commons APIs
nav_order: 10
---

# Create or update a controller
**Introduced 2.12**
{: .label .label-purple }

> Creates or updates a controller for a model. A model may be shared by multiple users. A controller sets rate limits for the number of [Predict API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/train-predict/predict/) calls users can make on the model. A controller consists of a set of rate limiters for different users.  

Author: ```suggestion
Use this API to create or update a controller for a model. A model may be shared by multiple users. A controller sets rate limits for the number of [Predict API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/train-predict/predict/) calls users can make on the model. A controller consists of a set of rate limiters for different users.  
```

The `user_rate_limiter` object contains an object for each user, specified by username. The user object contains the following fields.

Field | Data type | Description
:---  | :--- | :--- 
`limit` | Integer | The maximum number of times the user can call Predict API on the model per `unit` of time. By default, there is no limit on the number of Predict API calls. Once you set a limit, you cannot reset it to no limit. As an alternative, specify a high limit value and a small time unit, for example, 1 request per nanosecond.
`unit` | String | The unit of time for the rate limiter. Valid values are `DAYS`, `HOURS`, `MICROSECONDS`, `MILLISECONDS`, `MINUTES`, `NANOSECONDS`, and `SECONDS`.

#### Example request: Create controller

```json
POST _plugins/_ml/controllers/mtw-ZI0B_1JGmyB068C0
{
  "user_rate_limiter": {
    "user1": {
      "limit": 4,
      "unit": "MINUTES"
    },
    "user2": {
      "limit": 4,
      "unit": "MINUTES"
    }
  }
}
```
{% include copy-curl.html %}

#### Example response

```json
{
  "model_id": "mtw-ZI0B_1JGmyB068C0",
  "status": "CREATED"
}
```

> #### Example request: Update rate limit for one user

Author: ```suggestion
#### Example request: Update the rate limit for one user
```

To update the limit for `user1`, send a PUT request and specify the updated information:

```json
PUT _plugins/_ml/controllers/mtw-ZI0B_1JGmyB068C0
{
  "user_rate_limiter": {
    "user1": {
      "limit": 6,
      "unit": "MINUTES"
    }
  }
}
```
{% include copy-curl.html %}

This will update only the `user1` object, leaving all other user limits intact:

```json
{
  "model_id": "mtw-ZI0B_1JGmyB068C0",
  "user_rate_limiter": {
    "user1": {
      "limit": "6",
      "unit": "MINUTES"
    },
    "user2": {
      "limit": "4",
      "unit": "MINUTES"
    }
  }
}
```

#### Example response

```json
{
  "_index": ".plugins-ml-controller",
  "_id": "mtw-ZI0B_1JGmyB068C0",
  "_version": 2,
  "result": "updated",
  "forced_refresh": true,
  "_shards": {
    "total": 2,
    "successful": 2,
    "failed": 0
  },
  "_seq_no": 1,
  "_primary_term": 1
}
```

> #### Example request: Delete rate limit for one user

Author: ```suggestion
#### Example request: Delete the rate limit for one user
```
